src/v/cloud_storage/remote.h

```diff
@@ -274,10 +274,13 @@ class remote : public ss::peering_sharded_service<remote> {
     /// \param name The bucket to delete from
     /// \param parent The retry chain node to manage timeouts
     /// \param prefix Optional prefix to restrict listing of objects
+    /// \param item_filter Optional filter to apply to items before collecting
     ss::future<list_result> list_objects(
       const cloud_storage_clients::bucket_name& name,
       retry_chain_node& parent,
-      std::optional<cloud_storage_clients::object_key> prefix = std::nullopt);
+      std::optional<cloud_storage_clients::object_key> prefix = std::nullopt,
+      std::optional<cloud_storage_clients::client::item_filter> item_filter
```

Comment:
The trouble with using a predicate here is that the function still returns a `vector`, and the size of `list_bucket_item` is >32 bytes, so it only takes ~3000 topics for a listing of topic manifests to be allocating >128kB of memory.

Reply:
Would an intrusive list be appropriate here in place of vector?

Reply:
I haven't been able to address this in this PR, I would prefer to do this in a separate set of changes as this PR has grown quite large. Does that sound acceptable @jcsp ?